### ISSUE TITLE
Use new Tor control port features to solve all the huge problems

### ISSUE THREAD

**Issue by micahflee** [micahflee/onionshare#178]
Tor 0.2.7.1-alpha was [just released](https://blog.torproject.org/blog/tor-0271-alpha-released) and it includes some awesome new hidden service things.

> Major features (controller):
> * Add the ADD_ONION and DEL_ONION commands that allow the creation and management of hidden services via the controller. Closes ticket 6411.
> * New "GETINFO onions/current" and "GETINFO onions/detached" commands to get information about hidden services created via the controller. Part of ticket 6411.

I think using only the control port to create new hidden services, delete them when they're done, and learn the .onion address of them, will solve a ton of problems, including making #87, #130, and #153 all obsolete.

Obviously we wouldn't be able to deploy this until Tor 0.2.7.1 is included in Tor Browser. But I think I'd rather wait and use new cool Tor control port features than do a whole lot of complicated work.

**Comment by austin987** [comment]
Tor Browser bundle is now at 5.0.2, which ships with Tor 0.2.7.7.

**Comment by micahflee** [comment]
The 5.0.2 `tor-browser_en-US/Browser/TorBrowser/Docs/sources/versions` file says:

```
TOR_TAG=tor-0.2.6.10
TORLAUNCHER_TAG=0.2.7.7
```

So I think Tor is at 0.2.6.10, but Tor Launcher is later. However, the 5.5a2 version file says:

```
TOR_TAG=tor-0.2.7.2-alpha
TORLAUNCHER_TAG=0.2.7.7
```

So this is coming really soon. I'll start working on it.

**Comment by austin987** [comment]
You're right, I mixed up the two. But good to hear, thanks

**Comment by micahflee** [comment]
This has now been implemented in the master branch.

As long as the Tor available is at least `0.2.7.1` and the version of stem installed has [list_ephemeral_hidden_services](https://stem.torproject.org/api/control.html#stem.control.Controller.list_ephemeral_hidden_services) method implemented, onionshare will use ephemeral hidden services rather than the old way.

This means that for Linux, Tor Browser will still be required to run in the background until the `tor` and `python-stem` packages catch up, in whatever your distro is. In Windows and OSX I can bundle a new version of `python-stem`, but it will use the old way of hidden services unless the user is running Tor Browser 5.5.x (which is the current alpha) because it includes a newer version of `tor`.
